jstz_api/src/text_encoder.rs

```diff
@@ -1,71 +1,304 @@
+    }
+
+    fn custom_decode_utf16(input: &[u16]) -> String {
+        let input = input.into_iter().map(|&codepoint| codepoint);
+        let mut output: Vec<char> = Vec::new();
+        let mut utf16_decoder = char::decode_utf16(input);
+        while let Some(result) = utf16_decoder.next() {
+            match result {
+                Ok(s) => output.push(s),
+                Err(_) => break,
+            }
+        }
+        output.into_iter().collect()
+    }
+
+    fn encode_into(
+        source: &[u16],
+        destination: &mut Vec<u8>,
+    ) -> TextEncoderEncodeIntoResult {
+        //  1. Let read be 0
+        //  2. Let written be 0.
+        //  NB: read and written both correspond to the same value as UTF encoder can't fail
+        //      they are later set to `source` length
+        let read = source.len() as u128;
+        let written = read;
+
+        //  3. Let encoder be an instance of the UTF-8 encoder
+        //  NB: Noop, we only support UTF-8 encoder
+        //  4. Let unused be the I/O queue of scalar values << end-of-queue >>.
+        //  NB: The handler algorithm invoked below requires this argument, but it is not used by the UTF-8 encoder.
+        //  Therefore Noop
+        //  5. Convert src to an I/O queue of scalar values.
+        //  NB: here we just convert a u16 slice to a string slice using our custom decoder
+        //      it will stop after the first unpaired surrogate
+        let source = Self::custom_decode_utf16(source);
+        let input = source.as_str();
+        //  6. While true: ...
+        //  NB: this is performed by the rust `encoding` crate.
+        let () = UTF_8
+            .encode_to(input, EncoderTrap::Call(|_, _, _| false), destination)
+            .unwrap();
+        TextEncoderEncodeIntoResult { read, written }
+    }
+}
+#[derive(Default, Clone, Trace, Finalize)]
+pub struct TextEncoderClass;
+impl TextEncoderClass {
+    fn encode(_: &JsValue, args: &[JsValue], context: &mut Context) -> JsResult<JsValue> {
+        let arg = args.get_or_undefined(0).as_string().map(|x| x.as_slice());
+        let result = TextEncoder::encode(arg);
+        let byte_block = result.to_vec();
+        let array_buffer: JsArrayBuffer =
+            JsArrayBuffer::from_byte_block(byte_block, context)?;
+        let result_js = JsUint8Array::from_array_buffer(array_buffer, context)?;
+        Ok(result_js.into())
     }
-    fn btoa(_: &JsValue, args: &[JsValue], _: &mut Context) -> JsResult<JsValue> {
-        let data = args
+
+    fn encode_into(
+        _: &JsValue,
+        args: &[JsValue],
+        context: &mut Context,
+    ) -> JsResult<JsValue> {
+        let source = args
             .get_or_undefined(0)
             .as_string()
-            .ok_or_else(|| JsNativeError::typ().with_message("expected string"))?;
-        let result = TextEncoder::btoa(data)?;
-        Ok(result.into())
+            .map(|x| x.as_slice())
+            .unwrap();
```

Comment:
Avoid `.unwrap` unless the invariant is known. In this case you'd want to return a `JsError` (constructed using the `JsNativeError` API)

Reply:
done